File: gdb/testsuite/gdb.python/py-unwind.c
/* This test program is part of GDB, the GNU debugger.

   Copyright 2011-2014 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* This is the test program loaded into GDB by the py-unwind test.  */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

static void *
swap_value (void **location, void *new_value)
{
  void *old_value = *location;
  *location = new_value;
  return old_value;
}

#define MY_FRAME (__builtin_frame_address (0))

static void
break_backtrace ()
{
  /* Save outer frame address, then corrupt the unwind chain by
     setting the outer frame address in it to self.  This is
     ABI-specific: the first word of the frame contains previous frame
     address in amd64.  */
  void *outer_fp = swap_value ((void **)MY_FRAME, MY_FRAME);

  /* Verify the compiler allocates the first local variable one word
     below frame.  This is where test JIT reader expects to find the
     correct outer frame address.  */
  if (&outer_fp + 1 != (void **)MY_FRAME)
    {
      fprintf (stderr, "First variable should be allocated one word below "
               "the frame, got variable's address %p, frame at %p instead\n",
               &outer_fp, MY_FRAME);
      abort();
    }

  /* Now restore it so that we can return.  The test sets the
     breakpoint just before this happens, and GDB will not be able to
     show the backtrace without JIT reader.  */
  swap_value (MY_FRAME, outer_fp); /* break backtrace-broken */
}

static void
break_backtrace_caller ()
{
  break_backtrace ();
}

int
main (int argc, char *argv[])
{
  break_backtrace_caller ();
}

File: gdb/testsuite/gdb.python/py-unwind.py
# Copyright (C) 2013-2014 Free Software Foundation, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import gdb

def _read_word(address):
    return address.cast(_char_ptr_ptr_t).dereference()


def sniff (sniffer_info):
    "Sniffer written in Python."
    bp = sniffer_info.read_register(_amd64_rbp).cast(_char_ptr_t)
    try:
        if (_read_word(bp) == bp):
            # Found the frame that the test program fudged for us.
            # The correct BP for the outer frame has been saved one word
            # above, previous IP and SP are at the expected places
            previous_bp = _read_word(bp - 8)
            previous_ip = _read_word(bp + 8)
            previous_sp = bp + 16
            return (((_amd64_rbp, previous_bp),
                     (_amd64_rip, previous_ip),
                     (_amd64_rsp, previous_sp)),
                    (_amd64_rsp, _amd64_rip))

    except (gdb.error, RuntimeError):
        return None


_char_ptr_t = gdb.lookup_type("unsigned char").pointer()
_char_ptr_ptr_t = _char_ptr_t.pointer()
_uint_ptr_t = gdb.lookup_type("unsigned long long")
_amd64_rbp = 6
_amd64_rsp = 7
_amd64_rip = 16
gdb.frame_sniffers=[sniff]
print("Python script imported")
